core/src/main/scala/org/apache/spark/util/UninterruptibleThread.scala

```diff
@@ -92,11 +110,17 @@ private[spark] class UninterruptibleThread(
    * interrupted until it enters into the interruptible status.
    */
   override def interrupt(): Unit = {
```

Comment:
How about adding new flag to indicate if the thread is executed in `runUninterruptibly()` or not? If the thread doesn't run in `runUninterruptibly()`, we can interrupt the thread directly without accquiring the locks.

Reply:
Please check `uninterruptible`. This is the flag that is `true` for `runUninterruptibly()` and `false` otherwise:

```
  /**
   * Indicates if `this`  thread are in the uninterruptible status. If so, interrupting
   * "this" will be deferred until `this`  enters into the interruptible status.
   */
  @GuardedBy("uninterruptibleLock")
  private var uninterruptible = false
```

What concerns do you have with the current solution that works?

Reply:
My concern with the current solution is that it introduces unnecessary lock contention for the thread that run out of `runUninterruptibly()`. And I'm thinking if we could interrupt the thread more clearly, e.g.,

```scala
override def interrupt(): Unit = {
  if (lock {runInUninterruptibly}) {
    // interrupt as a uninterruptable thread
  } else {
   // interrupt as a normal thread
  }
}
```

Reply:
The solution does not introduce more lock contention. It reuses existing `uninterruptibleLock` and avoids holding it for the interruptible case. I don't see how your proposal would work without risk of interrupting uninterruptible case.